state-chain/pallets/cf-threshold-signature/src/lib.rs

```diff
@@ -367,28 +587,221 @@ pub mod pallet {
+		KeygenSuccessReported(T::ValidatorId),
+		/// A key handover participant has reported that keygen was successful \[validator_id\]
+		KeyHandoverSuccessReported(T::ValidatorId),
+		/// A keygen participant has reported that keygen has failed \[validator_id\]
+		KeygenFailureReported(T::ValidatorId),
+		/// A key handover participant has reported that keygen has failed \[validator_id\]
+		KeyHandoverFailureReported(T::ValidatorId),
+		/// Keygen was successful \[ceremony_id\]
+		KeygenSuccess(CeremonyId),
+		/// The key handover was successful
+		KeyHandoverSuccess {
+			ceremony_id: CeremonyId,
+		},
+		NoKeyHandover,
+		/// The new key was successfully used to sign.
+		KeygenVerificationSuccess {
+			agg_key: AggKeyFor<T, I>,
+		},
+		KeyHandoverVerificationSuccess {
+			agg_key: AggKeyFor<T, I>,
+		},
+		/// Verification of the new key has failed.
+		KeygenVerificationFailure {
+			keygen_ceremony_id: CeremonyId,
+		},
+		KeyHandoverVerificationFailure {
+			handover_ceremony_id: CeremonyId,
+		},
+		/// Keygen has failed \[ceremony_id\]
+		KeygenFailure(CeremonyId),
+		/// Keygen response timeout has occurred \[ceremony_id\]
+		KeygenResponseTimeout(CeremonyId),
+		KeyHandoverResponseTimeout {
+			ceremony_id: CeremonyId,
+		},
+		/// Keygen response timeout was updated \[new_timeout\]
+		KeygenResponseTimeoutUpdated {
+			new_timeout: BlockNumberFor<T>,
+		},
+		/// Key handover has failed
+		KeyHandoverFailure {
+			ceremony_id: CeremonyId,
+		},
+		/// The vault on chains associated with this key have all rotated
+		KeyRotationCompleted,
 	}
 
 	#[pallet::error]
 	pub enum Error<T, I = ()> {
 		/// The provided ceremony id is invalid.
-		InvalidCeremonyId,
+		InvalidThresholdSignatureCeremonyId,
+		/// An invalid keygen ceremony id
+		InvalidKeygenCeremonyId,
 		/// The provided threshold signature is invalid.
 		InvalidThresholdSignature,
 		/// The reporting party is not one of the signatories for this ceremony, or has already
 		/// responded.
-		InvalidRespondent,
+		InvalidThresholdSignatureRespondent,
+		/// An authority sent a response for a ceremony in which they weren't involved, or to which
+		/// they have already submitted a response.
+		InvalidKeygenRespondent,
 		/// The request Id is stale or not yet valid.
 		InvalidRequestId,
+		/// There is no threshold signature available
+		ThresholdSignatureUnavailable,
+		/// There is currently no rotation in progress for this key.
+		NoActiveRotation,
+		/// The requested call is invalid based on the current rotation state.
+		InvalidRotationStatus,
 	}
 
 	#[pallet::hooks]
 	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
-		fn on_initialize(current_block: BlockNumberFor<T>) -> frame_support::weights::Weight {
+		fn on_initialize(current_block: BlockNumberFor<T>) -> Weight {
+			let mut weight = T::DbWeight::get().reads(1);
+
+			// ====== 1. Process pending rotation taskes =======
```

Comment:
I think this is a good example of where we could break out the whole code block into a function in the submodule. Would allow us to test it independently too.